proposals/NNNN-reform-set.md

```diff
@@ -0,0 +1,222 @@
+# Reform SetAlgebra
+
+* Proposal: [SE-NNNN](NNNN-reform-set-algebra.md)
+* Authors: [Ben Cohen](https://github.com/airspeedswift)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN)
+
+## Introduction
+
+This proposal makes two changes to the `SetAlgebra` protocol, as well as corresponding changes to conforming types in the Standard Library (and Foundation):
+
+- amend `insert` to return the index to the current element in the set, not the element itself; and
+- split `SetAlgebra` into two protocols, for intensional and extensional sets.
+
+These are **source breaking** changes. The primary motivation for breaking source now is ensuring that the ABI and API allow for future types and language features – specifically move-only types.
+
+Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/)
+
+## Motivation
+
+### Preparation for move-only types
+
+As described in [SE-0232](https://github.com/apple/swift-evolution/blob/master/proposals/0232-remove-customization-points.md), requirements on a protocol where an element is returned, but not removed, from a collection will be a problem for collections that want to be able to hold move-only types (because returning the element means it would have to be moved out of the collection, so it can't remain in the collection).
+ There is one such method on the `SetAlgebra` protocol, `insert`:
+
+```swift
+/// Inserts the given element in the set if it is not already present.
+///
+/// If an element equal to `newMember` is already contained in the set, this
+/// method has no effect.
+///
+/// - Returns: `(true, newMember)` if `newMember` was not contained in the
+///   set. If an element equal to `newMember` was already contained in the
+///   set, the method returns `(false, oldMember)`, where `oldMember` is the
+///   element that was equal to `newMember`. In some cases, `oldMember` may
+///   be distinguishable from `newMember` by identity comparison or some
+///   other means.
+@discardableResult
+mutating func insert(_ newMember: Element) -> (inserted: Bool, memberAfterInsert: Element)
+```
+
+The operation returns the new value if an equal value wasn't already present; and _discards_ the new value and instead returns the existing old value if one was.
+
+This will be a problem for a set of move-only types. If the inserted type is move-only, it could not possibly be both inserted into the set _and_ returned.
+
+Contrast this with `Dictionary`'s `updateValue(_:forKey:)`, which replaces the existing value if there is one and returns that – something that would work fine for move-only types:
```

Comment:
`SetAlgebra` has an `update(with:)` requirement with similar behavior; it would probably make for a closer example.